File: workbench/_api/routes/j_lens.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel

from nnsightful.tools.j_lens import j_lens

from ..auth import require_user_email
from ..sse import stream_tool
from ..state import AppState, get_state

router = APIRouter()


class JLensRequest(BaseModel):
    model: str
    prompt: str
    topk: int = 5  # Number of top-k predictions per cell
    include_entropy: bool = True  # Whether to include entropy data


@router.post("/run")
async def run_j_lens(
    req: JLensRequest,
    state: AppState = Depends(get_state),
    user_email: str = Depends(require_user_email),
):
    """Run the Jacobian lens, streaming status until the data lands (see ``sse``)."""
    return stream_tool(
        state,
        j_lens,
        state[req.model],
        req.prompt,
        top_k=req.topk,
    )

File: workbench/_web/src/types/jlens.ts
/**
 * J-Lens Types
 *
 * The Jacobian lens (j-lens) reuses the logit-lens data format and widget, so
 * its data and UI-state types are re-exported from nnsightful's LogitLens types.
 * API-specific config types remain local.
 */

import type { LogitLensData, LogitLensUIState } from "nnsightful";

export type JLensData = LogitLensData;
export type JLensUIState = LogitLensUIState;

/**
 * J-Lens metadata
 */
export type JLensMeta = LogitLensData["meta"];

/**
 * J-Lens configuration data (mirrors Lens2 — same prompt/top-k/entropy knobs)
 */
export interface JLensConfigData {
    model: string;
    prompt: string;
    topk?: number; // Number of top-k predictions per cell (default: 5)
    includeEntropy?: boolean; // Whether to include entropy data (default: true)
    // Whether the run generates multiple tokens. Off = a single prefill pass
    // over the prompt (no generation); defaults off. When true the lens spans
    // prompt + generated tokens (JLensData.completion).
    generate?: boolean;
    // Generation length used when `generate` is on (≥2); defaults to 24. Ignored
    // when `generate` is off, where the run is a single pass (1 token).
    maxNewTokens?: number;
    // Sampling for the generation (only meaningful when maxNewTokens > 1). When
    // `sample` is false the run is greedy/deterministic and the knobs below are
    // ignored; when true they're forwarded to the backend's generate call.
    // `topK` here is the sampling top-k, distinct from `topk` (lens display).
    sample?: boolean;
    temperature?: number;
    topP?: number;
    topK?: number;
    // Persisted heatmap UI state (pinned trajectories, selection, layer
    // window, appearance) so the visualization restores across reloads.
    uiState?: LogitLensUIState;
}
